### PR TITLE
Improve `Taylor1{TaylorN{T}}` mixtures

### PR DESCRIPTION
**Pull request by lbenet**
This PR defines some specialized methods on `Taylor1{TaylorN{T}}`. This specialization is aimed to improve jet transport Taylor integrations.

### CODE REVIEW

**Review by PerezHz — COMMENTED** [review]
Many thanks @lbenet for adding these improvements to the handling of mixtures! Looks like this will really help TaylorIntegration.jl and NEOs.jl. I left a couple of comments about some minor stuff, but otherwise LGTM!

**Comment by PerezHz** [comment]
If I remember correctly, I think it was suggested in an offline discussion that it probably makes sense to add specialized evaluate methods for mixtures, is that still the case?

File: src/arithmetic.jl [hunk 1]
```diff
@@ -22,17 +22,18 @@ for T in (:Taylor1, :TaylorN)
     end
 end
 
-function ==(a::Taylor1{TaylorN{T}}, b::TaylorN{Taylor1{S}}) where {T,S}
-    R = promote_type(S,T)
-    return a == convert(Taylor1{TaylorN{R}},b)
+function ==(a::Taylor1{TaylorN{T}}, b::TaylorN{Taylor1{S}}) where {T, S}
+    R = promote_type(T, S)
+    return a == convert(Taylor1{TaylorN{S}}, b)
```
[the remark this answers is not in the document]
Comment (lbenet) [reply src/arithmetic.jl:27]: Great catch! Done!

File: src/arithmetic.jl [hunk 2]
```diff
@@ -309,22 +331,50 @@ for (f, fc) in ((:+, :(add!)), (:-, :(subst!)))
...
+
+        # function ($fc)(res::Taylor1{TaylorN{T}}, a::Taylor1{TaylorN{T}},
+        #         b::Taylor1{TaylorN{T}}, ordT::Int) where {T<:NumberNotSeries}
+        #     # Sanity
+        #     zero!(res, a, ordT)
+        #     @inbounds for ordQ in eachindex(a[ordT])
+        #         $(fc)(res[ordT], a[ordT], b[ordT], ordQ)
+        #     end
+        #     return nothing
+        # end
```
[the remark this answers is not in the document]
Comment (lbenet) [reply src/arithmetic.jl:377]: I think I wanted to have more specific methods, but something didn't work (ambiguities?). For the time being I'll leave them there, but if they are not needed, I'll remove them.

File: src/arithmetic.jl [hunk 3]
```diff
@@ -409,11 +458,28 @@ function *(a::HomogeneousPolynomial{T}, b::HomogeneousPolynomial{T}) where
...
 end
 
+function *(a::Taylor1{TaylorN{T}}, b::Taylor1{TaylorN{S}}) where
+        {T<:NumberNotSeries, S<:NumberNotSeries}
+    R = promote_type(T,S)
+    return *(convert(Taylor1{TaylorN{R}}, a), convert(Taylor1{TaylorN{R}}, b))
+end
+
+function *(a::Taylor1{TaylorN{T}}, b::Taylor1{TaylorN{T}}) where {T<:NumberNotSeries}
+    if (a.order != b.order) || any(get_order.(a[:]) .!= get_order.(b[:]))
```
[the remark this answers is not in the document]
Comment (lbenet) [reply src/arithmetic.jl:468]: Good point. implemented as you propose.

File: src/arithmetic.jl [hunk 4]
```diff
@@ -512,23 +592,28 @@ for T in (:Taylor1, :HomogeneousPolynomial, :TaylorN)
 
     @eval function /(a::$T{T}, b::T) where {T<:Number}
         @inbounds aux = a.coeffs[1] / b
-        v = Array{typeof(aux)}(undef, length(a.coeffs))
-        @__dot__ v = a.coeffs / b
-        return $T(v, a.order)
+        # v = Array{typeof(aux)}(undef, length(a.coeffs))
+        # @__dot__ v = a.coeffs / b
+        # return $T(v, a.order)
```
[the remark this answers is not in the document]
Comment (lbenet) [reply src/arithmetic.jl:597]: Yes, they should be eventually removed. I think I kept them to check performance. I shall keep this conversation open, to remember eventually remove them if not needed.

File: src/arithmetic.jl [hunk 5]
```diff
@@ -585,6 +675,22 @@ function /(a::TaylorN{T}, b::TaylorN{T}) where {T<:NumberNotSeriesN}
     return c
 end
 
+function /(a::Taylor1{TaylorN{T}}, b::Taylor1{TaylorN{T}}) where {T<:NumberNotSeries}
+    iszero(a) && !iszero(b) && return zero(a)
+    if (a.order != b.order) || any(get_order.(a[:]) .!= get_order.(b[:]))
```
[the remark this answers is not in the document]
Comment (lbenet) [reply src/arithmetic.jl:680]: Done!

Do you think indexing has an impact in performance? I mean, using a[k] vs a.coeffs[k+1]...

File: src/arithmetic.jl [hunk 6]
```diff
@@ -667,6 +774,40 @@ div!(v::Taylor1, b::NumberNotSeries, a::Taylor1, k::Int) =
     return nothing
 end
 
+# NOTE: Here `div!` *accumulates* the result of a / b in c[k] (k > 0)
```
[the remark this answers is not in the document]
Comment (lbenet) [reply src/arithmetic.jl:777]: Good catch! Changed.

File: src/auxiliary.jl [hunk 7]
```diff
@@ -264,6 +271,23 @@ function fixorder(a::HomogeneousPolynomial, b::HomogeneousPolynomial)
     return a, b
 end
 
+for T in (:HomogeneousPolynomial, :TaylorN)
+    @eval function fixorder(a::Taylor1{$T{T}}, b::Taylor1{$T{S}}) where
+            {T<:NumberNotSeries, S<:NumberNotSeries}
+        (a.order == b.order) && (all(get_order.(a[:]) .== get_order.(b[:]))) && return a, b
```
[the remark this answers is not in the document]
Comment (lbenet) [reply src/auxiliary.jl:277]: Done! Thanks!

**Comment by lbenet** [comment]
> If I remember correctly, I think it was suggested in an offline discussion that it probably makes sense to add specialized evaluate methods for mixtures, is that still the case?

Sorry for the long delay to get back to your comments... I have implemented most of your suggestions. About the comments, I think I left them there because I was testing improving performance, but either didn't get better results, or got some sort of ambiguities/errors. I left those conversations unresolved, to come back later to them.

File: src/arithmetic.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (PerezHz) [reply src/arithmetic.jl]: I have to say, I first thought indexing by a colon (i.e., `a[:]`) led to allocations, but after checking for a second time, it rather lowers to a `view`, thus it doesn't allocate. Therefore both solutions seem to me equivalent in principle in regards to performance, though `a.coeffs` could (maybe) be argued to be more generic

**Comment by PerezHz** [comment]
Many thanks for addressing the comments! I agree that probably there's room left to improve performance, probably we can merge this as a first step and then try to get further performance gains in future PRs. Thus, as it stands now, it's probably worth bumping at least the patch version?

**Comment by lbenet** [comment]
This follows the discussion from https://github.com/PerezHz/PlanetaryEphemeris.jl/pull/30#issuecomment-1915674000 

The following is the example used, using this branch
```julia
julia> using TaylorSeries

julia> x = Taylor1.([rand(3) for _ in 1:5, __ in 1:2])
5×2 Matrix{Taylor1{Float64}}:
    0.0991832863555554 + 0.6683889089955399 t + 0.7967925292561205 t² + 𝒪(t³)  …      0.489775369583631 + 0.1409952219760241 t + 0.7373398449922951 t² + 𝒪(t³)
...

julia> dq = TaylorSeries.set_variables("dq", order=2, numvars=2);

julia> x(3.04 + dq[1])
5×2 Matrix{TaylorN{Float64}}:
      9.494723408075362 + 5.512887486872753 dq₁ + 0.7967925292561205 dq₁² + 𝒪(‖x‖³)  …    7.732600755871539 + 4.624021479529178 dq₁ + 0.7373398449922951 dq₁² + 𝒪(‖x‖³)
...

julia> @code_warntype x(3.04 + dq[1])
MethodInstance for (::Matrix{Taylor1{Float64}})(::TaylorN{Float64})
  from (p::Array{Taylor1{T}})(x) where T<:Number @ TaylorSeries ~/.julia/dev/TaylorSeries/src/evaluate.jl:102
Static Parameters
  T = Float64
Arguments
  p::Matrix{Taylor1{Float64}}
  x::TaylorN{Float64}
Body::Any
1 ─ %1 = Base.broadcasted(TaylorSeries.evaluate, p, x)::Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{2}, Nothing, typeof(evaluate), Tuple{Matrix{Taylor1{Float64}}, TaylorN{Float64}}}
│   %2 = Base.materialize(%1)::Any
└──      return %2

julia> @which evaluate(x, 3.04 + dq[1])
evaluate(x::AbstractArray{Taylor1{T}}, δt::S) where {T<:Number, S<:Number}
     @ TaylorSeries ~/.julia/dev/TaylorSeries/src/evaluate.jl:42
```

The method quoted uses broadcasting, so I guess that is the problem.

**Comment by lbenet** [comment]
@PerezHz I think I have solved the issue you reported. I have checked that all the changes do not induce invalidations, and that tests pass in TaylorIntegration. In my opinion this is ready to be merged, except for tagging the new version; I am in favor of bumping a new minor version. Do you see anything that still has to be addressed?

**Comment by PerezHz** [comment]
@lbenet LGTM! I think there's further improvements to be made in the in-place methods like `mul!` for mixtures, but we can address this in future PRs (I have an idea regarding this in the works, will try to upload a very WIP proof-of-concept PR today)

**Comment by lbenet** [comment]
Great! So I'll bump the patch version, merge, and then release the new version!

**Pull request merged**